components/coding/profileV2/AcheivementComponent.tsx

```diff
@@ -1,62 +1,132 @@
 import React, { useEffect, useState } from "react";
 import ExpandableContainer from "../ExpandableContainer";
 import { PencilAltIcon } from "@heroicons/react/outline";
-import { IntroCourseUnit } from "../../../graphql/coding/userBadges/fetchUserBadges";
-import Link from "next/link";
+import { Data } from "../../../graphql/coding/userBadges/fetchUserBadges";
 import { User } from "../../../graphql/fetchUserProfile";
+import _ from "lodash";
 
 export type BadgesSectionProps = {
-  user: User;
   data: User;
 };
+export type userCodingBadge = {
+  id: number;
+};
+export type unitProps = {
+  unitTitle: string;
+  codingBadges: codingBadges[];
+};
+export type codingBadges = {
+  id: number;
+  title: string;
+  userCodingBadge: userCodingBadge[];
+  isAwarded: boolean;
+};
+export type unit = unitProps[];
+const AcheivementComponent = ({ data }) => {
+  const [transformedData, setTransformedData] = useState([]);
+  const [editMode, setEditMode] = useState(false);
+  const handleBadgeClick = (inputBadge: codingBadges) => {
+    setTransformedData((prev) => {
+      const updatedData = prev.map((unit: unitProps) => {
+        return {
+          ...unit,
+          codingBadges: unit.codingBadges.map((badge) => {
+            if (badge.id == inputBadge.id) {
+              return {
+                ...badge,
+                isAwarded: !badge.isAwarded,
+              };
+            } else {
+              return badge;
+            }
+          }),
+        };
+      });
 
-const AcheivementComponent = ({ user, data }) => {
-  const [units, setUnits] = useState<IntroCourseUnit[]>([]);
+      return updatedData;
+    });
+  };
+
+  const transformData = (data: Data) => {
```

Comment:
two things: 
- naming here is pretty generic, could end up being confusing down the line
- might be nice to move this to a helpers file since it doesn't have to be in here, in which case i'd definitely make the name more descriptive